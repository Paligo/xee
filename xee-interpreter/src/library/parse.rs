// https://www.w3.org/TR/xpath-functions-31/#parsing-and-serializing

use xee_xpath_macros::xpath_fn;

use crate::{context, error, interpreter::Interpreter, sequence, wrap_xpath_fn};

use super::StaticFunctionDescription;

#[xpath_fn("fn:parse-xml($arg as xs:string?) as document-node(element(*))?")]
fn parse_xml(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_string(interpreter.xot_mut(), None, arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:parse-xml-fragment($arg as xs:string?) as document-node()?")]
fn parse_xml_fragment(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_fragment_string(interpreter.xot_mut(), arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:serialize($arg as item()*) as xs:string")]
fn serialize1(interpreter: &mut Interpreter, arg: &sequence::Sequence) -> error::Result<String> {
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

#[xpath_fn("fn:serialize($arg as item()*, $params as item()?) as xs:string")]
fn serialize2(
    interpreter: &mut Interpreter,
    arg: &sequence::Sequence,
    _params: Option<sequence::Item>,
) -> error::Result<String> {
    // TODO: nothing is done with the parameters yet
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

pub(crate) fn static_function_descriptions() -> Vec<StaticFunctionDescription> {
    vec![
        wrap_xpath_fn!(parse_xml),
        wrap_xpath_fn!(parse_xml_fragment),
        wrap_xpath_fn!(serialize1),
        wrap_xpath_fn!(serialize2),
    ]
}
