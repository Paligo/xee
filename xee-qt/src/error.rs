use std::path::PathBuf;
use thiserror::Error;

use crate::outcome;
use crate::qt;

#[derive(Debug, Error)]
pub enum Error {
    #[error("Test failures {0} {1}")]
    TestFailures(PathBuf, outcome::TestSetOutcomes),
    #[error("catalog.xml cannot be found")]
    NoCatalogFound,
    #[error("File not found in catalog: {0}")]
    FileNotFoundInCatalog(PathBuf),
    #[error("Unknown environment reference")]
    UnknownEnvironmentReference(qt::EnvironmentRef),
    #[error("Cannot represent as XML")]
    CannotRepresentAsXml,
    #[error("Xee XPath error")]
    XeeXPath(#[from] xee_xpath::error::Error),
    #[error("Xot error")]
    Xot(#[from] xot::Error),
    #[error("IO error")]
    IO(#[from] std::io::Error),
    #[error("Var error")]
    VarError(#[from] std::env::VarError),
    #[error("Globset error")]
    GlobSet(#[from] globset::Error),
}

pub type Result<T> = std::result::Result<T, Error>;
