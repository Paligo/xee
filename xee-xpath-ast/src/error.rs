use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

#[derive(Debug, Clone, PartialEq, Error, Diagnostic)]
pub enum Error {
    /// Parse error.
    ///
    /// It is a static error if an expression is not a valid instance of the
    /// grammar defined in A.1 EBNF.
    #[error("Parse error")]
    #[diagnostic(code(XPST0003), help("Invalid XPath expression"))]
    ParseError {
        #[source_code]
        src: String,
        #[label("Could not parse beyond this")]
        span: SourceSpan,
    },
}
