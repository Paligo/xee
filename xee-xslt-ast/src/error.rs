use crate::ast_core::Span;
use crate::value_template;

#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct XmlName {
    pub namespace: String,
    pub local: String,
}

#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Error {
    // We didn't get the element we expect
    Unexpected,
    // Expected attribute of name, not found (element span)
    AttributeExpected {
        name: XmlName,
        span: Span,
    },
    // Did not expect attribute of name (attribute span)
    AttributeUnexpected {
        name: XmlName,
        span: Span,
        message: String,
    },
    // The value of the an attribute was invalid
    Invalid {
        value: String,
        span: Span,
    },

    UnexpectedSequenceConstructor,
    InvalidInstruction {
        span: Span,
    },
    InvalidValueTemplate {
        span: Span,
    },

    MissingSpan,
    InvalidEqName {
        value: String,
        span: Span,
    },

    ElementMissing {
        span: Span,
    },

    UnexpectedElement {
        name: XmlName,
        span: Span,
    },
    ExpectedElementNotFound {
        expected: XmlName,
        span: Span,
    },
    /// An internal error; this indicates a bug as some invariant in the
    /// code wasn't met.
    Internal(&'static str),
    XPath(xee_xpath_ast::ParserError),
}

pub type Result<T> = std::result::Result<T, Error>;

impl From<xee_xpath_ast::ParserError> for Error {
    fn from(error: xee_xpath_ast::ParserError) -> Self {
        Self::XPath(error)
    }
}

impl From<value_template::Error> for Error {
    fn from(error: value_template::Error) -> Self {
        match error {
            value_template::Error::UnescapedCurly { span, .. } => {
                Self::InvalidValueTemplate { span }
            }
            value_template::Error::IllegalSlice => Self::Internal("Illegal slice"),
            value_template::Error::XPath(e) => Self::XPath(e),
        }
    }
}

impl Error {
    pub(crate) fn is_unexpected(&self) -> bool {
        matches!(self, Self::Unexpected)
    }
}
