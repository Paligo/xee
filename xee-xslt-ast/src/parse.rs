use ahash::HashMap;
use chumsky::util::MaybeRef;
use chumsky::{extra::Full, input::ValueInput, prelude::*};
use std::borrow::Cow;
use xee_xpath_ast::Namespaces;
use xot::{Node, Xot};

use crate::ast_core as ast;

pub(crate) struct State<'a> {
    pub(crate) namespaces: Cow<'a, Namespaces<'a>>,
}

type Extra<'a> = Full<ParserError, State<'a>, ()>;

pub(crate) type BoxedParser<'a, I, T> = Boxed<'a, 'a, I, T, Extra<'a>>;

pub type Span = SimpleSpan;

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Token<'a> {
    ElementStart(Name<'a>, HashMap<Name<'a>, &'a str>),
    ElementEnd(Name<'a>),
    Text(&'a str),
    Comment(&'a str),
    ProcessingInstruction(&'a str, Option<&'a str>),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Name<'a> {
    namespace: &'a str,
    localname: &'a str,
}

impl<'a> From<(&'a str, &'a str)> for Name<'a> {
    fn from((localname, namespace): (&'a str, &'a str)) -> Self {
        Self {
            namespace,
            localname,
        }
    }
}

#[cfg_attr(test, derive(serde::Serialize))]
pub enum ParserError {
    ExpectedFound { span: Span },
    MyError,
    XPath(xee_xpath_ast::ParserError),
}

impl From<xee_xpath_ast::ParserError> for ParserError {
    fn from(e: xee_xpath_ast::ParserError) -> Self {
        Self::XPath(e)
    }
}

impl<'a, I> chumsky::error::Error<'a, I> for ParserError
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    // we don't do anything with expected and found, instead just retaining
    // the span. This is because these contain tokens with a lifetime, and
    // having a lifetime for the ParserError turns out open up a world of trouble
    // as soon as we want to build on it in the XSLT parser. We also don't
    // have a good way to turn a logos token into a human-readable string, so
    // we couldn't really construct good error messages anyway.
    fn expected_found<E: IntoIterator<Item = Option<MaybeRef<'a, Token<'a>>>>>(
        _expected: E,
        _found: Option<MaybeRef<'a, Token<'a>>>,
        span: Span,
    ) -> Self {
        Self::ExpectedFound { span }
    }

    fn merge(self, other: Self) -> Self {
        match (self, other) {
            (
                ParserError::ExpectedFound { span: span_a },
                ParserError::ExpectedFound { span: _ },
            ) => ParserError::ExpectedFound { span: span_a },
            (ParserError::ExpectedFound { .. }, a) => a,
            (a, ParserError::ExpectedFound { .. }) => a,
            (a, _) => a,
        }
    }
}

struct TokenizedTraverse<'a, T: Iterator<Item = xot::NodeEdge>> {
    xot: &'a Xot,
    traverse: T,
}

impl<'a, T: Iterator<Item = xot::NodeEdge>> Iterator for TokenizedTraverse<'a, T> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        let edge = self.traverse.next()?;
        Some(match edge {
            xot::NodeEdge::Start(node) => match self.xot.value(node) {
                xot::Value::Element(e) => {
                    let name: Name = self.xot.name_ns_str(e.name()).into();
                    let attributes = e
                        .attributes()
                        .iter()
                        .map(|(name, value)| {
                            let name: Name = self.xot.name_ns_str(*name).into();
                            (name, value.as_ref())
                        })
                        .collect::<HashMap<_, _>>();
                    Token::ElementStart(name, attributes)
                }
                xot::Value::Text(text) => Token::Text(text.get()),
                xot::Value::Comment(comment) => Token::Comment(comment.get()),
                xot::Value::ProcessingInstruction(pi) => {
                    Token::ProcessingInstruction(pi.target(), pi.data())
                }
                xot::Value::Root => self.next()?,
            },
            xot::NodeEdge::End(node) => {
                if let xot::Value::Element(e) = self.xot.value(node) {
                    let name: Name = self.xot.name_ns_str(e.name()).into();
                    Token::ElementEnd(name)
                } else {
                    self.next()?
                }
            }
        })
    }
}

fn tokenize(xot: &Xot, node: Node) -> TokenizedTraverse<impl Iterator<Item = xot::NodeEdge> + '_> {
    TokenizedTraverse {
        xot,
        traverse: xot.traverse(node),
    }
}

fn parser<'a, I>() -> BoxedParser<'a, I, ast::If>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    let if_ = select! {
        Token::ElementStart(Name { namespace: "", localname: "if"}, attrs) => attrs,
    }
    .boxed();

    let text = select! {
        Token::Text(s) => s,
    }
    .boxed();

    let sequence_constructor = text
        .map(|text| ast::SequenceConstructorItem::Text(text.to_string()))
        .boxed();

    if_.then(sequence_constructor)
        .try_map_with_state(|(attributes, content), _span, state: &mut State| {
            let name = Name {
                namespace: "",
                localname: "test",
            };
            let test = attributes.get(&name).unwrap();
            let namespaces = state.namespaces.as_ref();

            let test = xee_xpath_ast::ast::XPath::parse(test, namespaces, &[])?;
            Ok(ast::If {
                test,
                content: vec![content],
            })
        })
        .then_ignore(just(Token::ElementEnd(Name {
            namespace: "",
            localname: "if",
        })))
        .boxed()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::input::Stream;
    use insta::assert_ron_snapshot;

    #[test]
    fn test_tokenize() {
        let mut xot = Xot::new();
        let doc = xot.parse(r#"<if test="true()">Hello</if>"#).unwrap();
        let tokens = tokenize(&xot, doc).collect::<Vec<_>>();
        assert_ron_snapshot!(tokens);
    }

    #[test]
    fn test_simple_parse_if() {
        let mut xot = Xot::new();
        let doc = xot.parse(r#"<if test="true()">Hello</if>"#).unwrap();
        let tokens = tokenize(&xot, doc);
        let stream = Stream::from_iter(tokens);
        let namespaces = Namespaces::default();
        let mut state = State {
            namespaces: Cow::Owned(namespaces),
        };
        assert_ron_snapshot!(parser().parse_with_state(stream, &mut state).into_result());
    }
}
