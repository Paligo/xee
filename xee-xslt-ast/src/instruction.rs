use xot::{NameId, Xot};

use crate::ast_core as ast;
use crate::error::Error;
use crate::names::SequenceConstructorName;
use crate::parse::Element;

pub(crate) trait InstructionParser: Sized + Into<ast::SequenceConstructorItem> {
    fn parse(element: &Element) -> Result<ast::SequenceConstructorItem, Error> {
        let ast = Self::parse_ast(element)?;
        ast.validate(element)?;
        Ok(ast.into())
    }

    fn validate(&self, _element: &Element) -> Result<(), Error> {
        Ok(())
    }

    fn parse_ast(element: &Element) -> Result<Self, Error>;
}

impl InstructionParser for ast::SequenceConstructorItem {
    fn parse_ast(element: &Element) -> Result<ast::SequenceConstructorItem, Error> {
        let sname = element
            .names
            .sequence_constructor_name(element.element.name());

        if let Some(sname) = sname {
            // parse a known sequence constructor instruction
            match sname {
                SequenceConstructorName::Copy => ast::Copy::parse(element),
                SequenceConstructorName::If => ast::If::parse(element),
                SequenceConstructorName::Variable => ast::Variable::parse(element),
            }
        } else {
            let ns = element.xot.namespace_for_name(element.element.name());
            if ns == element.names.xsl_ns {
                // we have an unknown xsl instruction, fail with error
                Err(Error::InvalidInstruction { span: element.span })
            } else {
                // we parse the literal element
                ast::ElementNode::parse(element)
            }
        }
    }
}

impl InstructionParser for ast::ElementNode {
    fn parse_ast(element: &Element) -> Result<ast::ElementNode, Error> {
        Ok(ast::ElementNode {
            name: to_name(element.xot, element.element.name()),

            standard: element.xsl_standard()?,
            span: element.span,
        })
    }
}

fn to_name(xot: &Xot, name: NameId) -> ast::Name {
    let (local, namespace) = xot.name_ns_str(name);
    ast::Name {
        namespace: namespace.to_string(),
        local: local.to_string(),
    }
}

// impl InstructionParser for ast::Assert {
//     fn parse_ast(element: &Element) -> Result<Self, Error> {
//         let names = element.names;
//         Ok(ast::Assert {
//             test: element.required(names.test, |s, span| element.xpath(s, span))?,
//             select: element.optional(names.select, |s, span| element.xpath(s, span))?,
//             error_code: element
//                 .optional(names.error_code, |s, span| element.attribute_value(s, span))?,
//             content: element.sequence_constructor()?,

//             standard: element.standard()?,
//             span: element.span,
//         })
//     }
// }

impl InstructionParser for ast::Copy {
    fn parse_ast(element: &Element) -> Result<Self, Error> {
        let names = element.names;
        Ok(ast::Copy {
            select: element.optional(names.select, |s, span| element.xpath(s, span))?,
            copy_namespaces: element.boolean(names.copy_namespaces, true)?,
            inherit_namespaces: element.boolean(names.inherit_namespaces, true)?,
            use_attribute_sets: element.optional(names.use_attribute_sets, Element::eqnames)?,
            type_: element.optional(names.as_, Element::eqname)?,
            validation: element
                .optional(names.validation, Element::validation)?
                // TODO: should depend on global validation attribute
                .unwrap_or(ast::Validation::Strip),
            content: element.sequence_constructor()?,
            standard: element.standard()?,
            span: element.span,
        })
    }
}

// impl InstructionParser for ast::Fallback {
//     fn parse_ast(element: &Element) -> Result<Self, Error> {
//         let parser = element.xslt_parser;
//         let content =
//         Ok(ast::Fallback {
//             content: parser.parse_sequence_constructor(element.node)?;
//             span: element.span,
//         })
//     }
// }

impl InstructionParser for ast::If {
    fn parse_ast(element: &Element) -> Result<Self, Error> {
        let names = element.names;
        Ok(ast::If {
            test: element.required(names.test, |s, span| element.xpath(s, span))?,
            content: element.sequence_constructor()?,
            standard: element.standard()?,
            span: element.span,
        })
    }
}

impl InstructionParser for ast::Variable {
    fn parse_ast(element: &Element) -> Result<Self, Error> {
        let names = element.names;

        // This is a rule somewhere, but not sure whether it's correct;
        // can visibility be absent or is there a default visibility?
        // let visibility = visibility.unwrap_or(if static_ {
        //     ast::VisibilityWithAbstract::Private
        // } else {
        //     ast::VisibilityWithAbstract::Public
        // });

        Ok(ast::Variable {
            name: element.required(names.name, Element::eqname)?,
            select: element.optional(names.select, |s, span| element.xpath(s, span))?,
            as_: element.optional(names.as_, |s, span| element.sequence_type(s, span))?,
            static_: element.boolean(names.static_, false)?,
            visibility: element.optional(names.visibility, Element::visibility_with_abstract)?,
            content: element.sequence_constructor()?,
            standard: element.standard()?,
            span: element.span,
        })
    }

    fn validate(&self, element: &Element) -> Result<(), Error> {
        if self.visibility == Some(ast::VisibilityWithAbstract::Abstract) && self.select.is_some() {
            return Err(element.attribute_unexpected(
                element.names.select,
                "select attribute is not allowed when visibility is abstract",
            ));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {

    use crate::{names::Names, parse::XsltParser};

    use super::*;
    use insta::assert_ron_snapshot;
    use xee_xpath_ast::Namespaces;

    fn parse(s: &str) -> Result<ast::SequenceConstructorItem, Error> {
        let mut xot = Xot::new();
        let names = Names::new(&mut xot);
        let namespaces = Namespaces::default();

        let (node, span_info) = xot.parse_with_span_info(s).unwrap();
        let node = xot.document_element(node).unwrap();
        let parser = XsltParser::new(&xot, &names, &span_info, namespaces);
        parser.parse(node)
    }

    #[test]
    fn test_if() {
        assert_ron_snapshot!(parse(
            r#"<xsl:if xmlns:xsl="http://www.w3.org/1999/XSL/Transform" test="true()">Hello</xsl:if>"#
        ));
    }

    #[test]
    fn test_variable() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" select="true()">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_missing_required() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" select="true()">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_broken_xpath() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" select="let $x := 1">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_sequence_type() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" as="xs:string" select="true()">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_boolean_default_no_with_explicit_yes() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" static="yes" as="xs:string" select="true()">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_variable_visibility() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" visibility="public">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_variable_visibility_abstract_with_select_is_error() {
        assert_ron_snapshot!(parse(
            r#"<xsl:variable xmlns:xsl="http://www.w3.org/1999/XSL/Transform" name="foo" visibility="abstract" select="true()">Hello</xsl:variable>"#
        ));
    }

    #[test]
    fn test_copy() {
        assert_ron_snapshot!(parse(
            r#"<xsl:copy xmlns:xsl="http://www.w3.org/1999/XSL/Transform" select="true()" copy-namespaces="no" inherit-namespaces="no" validation="strict">Hello</xsl:copy>"#
        ));
    }

    #[test]
    fn test_eqnames() {
        assert_ron_snapshot!(parse(
            r#"<xsl:copy xmlns:xsl="http://www.w3.org/1999/XSL/Transform" use-attribute-sets="foo bar baz">Hello</xsl:copy>"#
        ));
    }

    #[test]
    fn test_nested_if() {
        assert_ron_snapshot!(parse(
            r#"<xsl:if xmlns:xsl="http://www.w3.org/1999/XSL/Transform" test="true()"><xsl:if test="true()">Hello</xsl:if></xsl:if>"#
        ));
    }

    #[test]
    fn test_if_with_standard_attribute() {
        assert_ron_snapshot!(parse(
            r#"<xsl:if xmlns:xsl="http://www.w3.org/1999/XSL/Transform" test="true()" expand-text="yes">Hello</xsl:if>"#
        ));
    }

    #[test]
    fn test_literal_result_element() {
        assert_ron_snapshot!(parse(r#"<foo/>"#));
    }

    #[test]
    fn test_literal_result_element_with_standard_attribute() {
        assert_ron_snapshot!(parse(
            r#"<foo xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xsl:expand-text="yes"/>"#
        ));
    }
}
