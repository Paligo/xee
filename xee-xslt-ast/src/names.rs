use std::collections::BTreeMap;
use std::str::FromStr;

use strum::VariantNames;
use xot::{NameId, NamespaceId, Xot};

use crate::ast_core::{self as ast, DeclarationName, SequenceConstructorName};
use crate::combinator::ElementError;
use crate::element::Element;
use crate::instruction::{DeclarationParser, SequenceConstructorParser};

impl SequenceConstructorName {
    pub(crate) fn parse(
        &self,
        element: &Element,
    ) -> Result<ast::SequenceConstructorItem, ElementError> {
        match self {
            SequenceConstructorName::ApplyImports => {
                ast::ApplyImports::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::AnalyzeString => {
                ast::AnalyzeString::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::ApplyTemplates => {
                ast::ApplyTemplates::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::Assert => {
                ast::Assert::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::Attribute => {
                ast::Attribute::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::Copy => ast::Copy::parse_sequence_constructor_item(element),
            SequenceConstructorName::CopyOf => {
                ast::CopyOf::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::If => ast::If::parse_sequence_constructor_item(element),
            SequenceConstructorName::Variable => {
                ast::Variable::parse_sequence_constructor_item(element)
            }
            SequenceConstructorName::Fallback => {
                ast::Fallback::parse_sequence_constructor_item(element)
            }
            _ => {
                unimplemented!()
            }
        }
    }

    fn names(xot: &mut Xot, xsl_ns: xot::NamespaceId) -> BTreeMap<NameId, SequenceConstructorName> {
        let mut sequence_constructor_names = BTreeMap::new();
        for variant_name in Self::VARIANTS {
            let name = xot.add_name_ns(variant_name, xsl_ns);
            let constructor = SequenceConstructorName::from_str(variant_name).unwrap();
            sequence_constructor_names.insert(name, constructor);
        }
        sequence_constructor_names
    }
}

impl DeclarationName {
    pub(crate) fn parse(&self, element: &Element) -> Result<ast::Declaration, ElementError> {
        match self {
            DeclarationName::Accumulator => ast::Accumulator::parse_declaration(element),
            _ => {
                unimplemented!()
            }
        }
    }

    fn names(xot: &mut Xot, xsl_ns: xot::NamespaceId) -> BTreeMap<NameId, DeclarationName> {
        let mut declaration_names = BTreeMap::new();
        for variant_name in Self::VARIANTS {
            let name = xot.add_name_ns(variant_name, xsl_ns);
            let constructor = DeclarationName::from_str(variant_name).unwrap();
            declaration_names.insert(name, constructor);
        }
        declaration_names
    }
}

pub(crate) struct Names {
    pub(crate) xsl_ns: NamespaceId,

    pub(crate) sequence_constructor_names: BTreeMap<NameId, SequenceConstructorName>,
    pub(crate) declaration_names: BTreeMap<NameId, DeclarationName>,

    // XSL elements
    pub(crate) xsl_accumulator_rule: xot::NameId,
    pub(crate) xsl_attribute: xot::NameId,
    pub(crate) xsl_fallback: xot::NameId,
    pub(crate) xsl_matching_substring: xot::NameId,
    pub(crate) xsl_non_matching_substring: xot::NameId,
    pub(crate) xsl_sort: xot::NameId,
    pub(crate) xsl_transform: xot::NameId,
    pub(crate) xsl_with_param: xot::NameId,

    // attributes on XSLT elements
    pub(crate) as_: xot::NameId,
    pub(crate) case_order: xot::NameId,
    pub(crate) component: xot::NameId,
    pub(crate) collation: xot::NameId,
    pub(crate) copy_accumulators: xot::NameId,
    pub(crate) copy_namespaces: xot::NameId,
    pub(crate) data_type: xot::NameId,
    pub(crate) error_code: xot::NameId,
    pub(crate) extension_element_prefixes: xot::NameId,
    pub(crate) flags: xot::NameId,
    pub(crate) id: xot::NameId,
    pub(crate) inherit_namespaces: xot::NameId,
    pub(crate) initial_value: xot::NameId,
    pub(crate) input_type_annotations: xot::NameId,
    pub(crate) lang: xot::NameId,
    pub(crate) match_: xot::NameId,
    pub(crate) mode: xot::NameId,
    pub(crate) name: xot::NameId,
    pub(crate) names: xot::NameId,
    pub(crate) namespace: xot::NameId,
    pub(crate) order: xot::NameId,
    pub(crate) phase: xot::NameId,
    pub(crate) regex: xot::NameId,
    pub(crate) select: xot::NameId,
    pub(crate) separator: xot::NameId,
    pub(crate) stable: xot::NameId,
    pub(crate) static_: xot::NameId,
    pub(crate) streamable: xot::NameId,
    pub(crate) test: xot::NameId,
    pub(crate) type_: xot::NameId,
    pub(crate) tunnel: xot::NameId,
    pub(crate) use_attribute_sets: xot::NameId,
    pub(crate) validation: xot::NameId,
    pub(crate) visibility: xot::NameId,

    // standard attributes on XSLT elements
    pub(crate) standard: StandardNames,
    // standard attributes on literal result elements
    pub(crate) xsl_standard: StandardNames,
}

pub(crate) struct StandardNames {
    pub(crate) default_collation: xot::NameId,
    pub(crate) default_mode: xot::NameId,
    pub(crate) default_validation: xot::NameId,
    pub(crate) exclude_result_prefixes: xot::NameId,
    pub(crate) expand_text: xot::NameId,
    pub(crate) extension_element_prefixes: xot::NameId,
    pub(crate) use_when: xot::NameId,
    pub(crate) version: xot::NameId,
    pub(crate) xpath_default_namespace: xot::NameId,
}

impl StandardNames {
    fn no_ns(xot: &mut Xot) -> Self {
        Self {
            default_collation: xot.add_name("default-collation"),
            default_mode: xot.add_name("default-mode"),
            default_validation: xot.add_name("default-validation"),
            exclude_result_prefixes: xot.add_name("exclude-result-prefixes"),
            expand_text: xot.add_name("expand-text"),
            extension_element_prefixes: xot.add_name("extension-element-prefixes"),
            use_when: xot.add_name("use-when"),
            version: xot.add_name("version"),
            xpath_default_namespace: xot.add_name("xpath-default-namespace"),
        }
    }

    fn xsl(xot: &mut Xot, xsl_ns: NamespaceId) -> Self {
        Self {
            default_collation: xot.add_name_ns("default-collation", xsl_ns),
            default_mode: xot.add_name_ns("default-mode", xsl_ns),
            default_validation: xot.add_name_ns("default-validation", xsl_ns),
            exclude_result_prefixes: xot.add_name_ns("exclude-result-prefixes", xsl_ns),
            expand_text: xot.add_name_ns("expand-text", xsl_ns),
            extension_element_prefixes: xot.add_name_ns("extension-element-prefixes", xsl_ns),
            use_when: xot.add_name_ns("use-when", xsl_ns),
            version: xot.add_name_ns("version", xsl_ns),
            xpath_default_namespace: xot.add_name_ns("xpath-default-namespace", xsl_ns),
        }
    }
}

impl Names {
    pub(crate) fn new(xot: &mut Xot) -> Self {
        let xsl_ns = xot.add_namespace("http://www.w3.org/1999/XSL/Transform");

        Self {
            xsl_ns,

            sequence_constructor_names: SequenceConstructorName::names(xot, xsl_ns),
            declaration_names: DeclarationName::names(xot, xsl_ns),

            xsl_accumulator_rule: xot.add_name_ns("accumulator-rule", xsl_ns),
            xsl_attribute: xot.add_name_ns("attribute", xsl_ns),
            xsl_fallback: xot.add_name_ns("fallback", xsl_ns),
            xsl_matching_substring: xot.add_name_ns("matching-substring", xsl_ns),
            xsl_non_matching_substring: xot.add_name_ns("non-matching-substring", xsl_ns),
            xsl_sort: xot.add_name_ns("sort", xsl_ns),
            xsl_transform: xot.add_name_ns("transform", xsl_ns),
            xsl_with_param: xot.add_name_ns("with-param", xsl_ns),

            as_: xot.add_name("as"),
            case_order: xot.add_name("case-order"),
            collation: xot.add_name("collation"),
            component: xot.add_name("component"),
            copy_accumulators: xot.add_name("copy-accumulators"),
            copy_namespaces: xot.add_name("copy-namespaces"),
            data_type: xot.add_name("data-type"),
            error_code: xot.add_name("error-code"),
            extension_element_prefixes: xot.add_name("extension-element-prefixes"),
            flags: xot.add_name("flags"),
            id: xot.add_name("id"),
            inherit_namespaces: xot.add_name("inherit-namespaces"),
            initial_value: xot.add_name("initial-value"),
            input_type_annotations: xot.add_name("input-type-annotations"),
            lang: xot.add_name("language"),
            match_: xot.add_name("match"),
            mode: xot.add_name("mode"),
            name: xot.add_name("name"),
            names: xot.add_name("names"),
            namespace: xot.add_name("namespace"),
            order: xot.add_name("order"),
            phase: xot.add_name("phase"),
            regex: xot.add_name("regex"),
            select: xot.add_name("select"),
            separator: xot.add_name("separator"),
            stable: xot.add_name("stable"),
            static_: xot.add_name("static"),
            streamable: xot.add_name("streamable"),
            test: xot.add_name("test"),
            tunnel: xot.add_name("tunnel"),
            type_: xot.add_name("type"),
            use_attribute_sets: xot.add_name("use-attribute-sets"),
            validation: xot.add_name("validation"),
            visibility: xot.add_name("visibility"),

            // standard attributes
            standard: StandardNames::no_ns(xot),
            // standard attributes on literal result elements
            xsl_standard: StandardNames::xsl(xot, xsl_ns),
        }
    }

    pub(crate) fn sequence_constructor_name(
        &self,
        name: NameId,
    ) -> Option<SequenceConstructorName> {
        self.sequence_constructor_names.get(&name).copied()
    }

    pub(crate) fn declaration_name(&self, name: NameId) -> Option<DeclarationName> {
        self.declaration_names.get(&name).copied()
    }
}
