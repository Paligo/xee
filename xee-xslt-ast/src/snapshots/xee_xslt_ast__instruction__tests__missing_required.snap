---
source: xee-xslt-ast/src/instruction.rs
expression: "parse(r#\"<xsl:variable xmlns:xsl=\"http://www.w3.org/1999/XSL/Transform\" select=\"true()\">Hello</xsl:variable>\"#)"
---
Err(AttributeExpected(
  namespace: "",
  local: "name",
  span: Span(
    start: 1,
    end: 13,
  ),
))
